Move MSLayerKeeper* objects from thread-local into EventSetup

makortel opened the pull request:
#### PR description:

This PR is part of https://github.com/cms-sw/cmssw/issues/31061 and focuses on the `MultipleScatteringParametrisation` helper. That class relied an old hack to cache EventSetup data
https://github.com/cms-sw/cmssw/blob/919a7477d148ff35ff5d5174e117d27227c87284/RecoTracker/TkMSParametrization/src/MultipleScatteringParametrisation.cc#L45-L49
that this PR suggests to finally fix properly by moving the `MSLayersKeeperX0DetLayer`, `MSLayersKeeperX0AtEta`, and `MSLayersKeeperX0Averaged` from thread-locals into EventSetup itself. In the EventSetup they are abstracted into `MultipleScatteringParametrisationMaker` class that is then used to make `MultipleScatteringParametrisation` objects. The corresponding ESProducer is added into the `customizeHLTforCMSSW.py` and in various cff fragments within reconstruction configuration (I can't exclude missing some places).

This PR also completes the esConsumes migration for touched helper classes where that was feasible.

As in https://github.com/cms-sw/cmssw/pull/35081, despite of being part of #31061 this PR adds some new calls to the deprecated `EventSetupRecord::get()` function without `esConsumes()`. All these cases have some relation to `TrackingRegion` classes, and will be addressed in a subsequent PR (I have the work done already, but I though they would be easier to review separately).

Resolves https://github.com/makortel/framework/issues/251

#### PR validation:

Tested on earlier IB (on top of https://github.com/cms-sw/cmssw/pull/35081) that limited matrix runs. After rebase checked that the code compiles.

makortel committed:
Add MultipleScatteringParametrisationMaker ESProduct that makes MultipleScatteringParametrisation objects

In order to store MSLayerKeeper* objects in EventSetup instead of thread local objects.

makortel committed:
Migrate code to use MultipleScatteringParametrisationMaker

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35269/25261

- This PR adds an extra 104KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HLTrigger/Configuration/python/customizeHLTforCMSSW.py  modified in PR(s): #34978, #34788
  - File  RecoTracker/IterativeTracking/python/MixedTripletStep_cff.py  modified in PR(s): #35194

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

- HLTrigger/Configuration (**hlt**)
- RecoHI/HiTracking (**reconstruction**)
- RecoPixelVertexing/PixelLowPtUtilities (**reconstruction**)
- RecoPixelVertexing/PixelTrackFitting (**reconstruction**)
- RecoPixelVertexing/PixelTriplets (**reconstruction**)
- RecoTracker/IterativeTracking (**reconstruction**)
- RecoTracker/TkHitPairs (**reconstruction**)
- RecoTracker/TkMSParametrization (**reconstruction**)
- RecoTracker/TkSeedGenerator (**reconstruction**)
- RecoTracker/TkTrackingRegions (**reconstruction**)


@Martin-Grunewald, @jpata, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@mtosi, @felicepantaleo, @GiacomoSguazzoni, @JanFSchulte, @jazzitup, @VinInn, @Martin-Grunewald, @yenjie, @ebrondol, @rovere, @kurtejung, @gpetruc, @mmusich, @mandrenguyen, @dgulhan, @yetkinyilmaz this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test

makortel commented on RecoTracker/TkMSParametrization/plugins/MultipleScatteringParametrisationMakerESProducer.cc:
```diff
@@ -0,0 +1,47 @@
+  // Using NavigationSchoolRecord here sounds a bit confusing, that
+  // that depends only on TrackerRecoGeometryRecord and
+  // IdealMagneticFieldRecord
+  std::unique_ptr<MultipleScatteringParametrisationMaker> produce(const NavigationSchoolRecord& iRecord);
```
Note that I blatantly reused the `NavigationSchoolRecord` for this ESProduct, as that record depends exactly on the two records (`TrackerRecoGeometryRecord` and `IdealMagneticFieldRecord`) needed for the `MultipleScatteringParametrisation`). If that is considered too confusing, I can create a new record (or put it to some other existing record like `CkfComponentsRecord`, but those may have smaller IOVs than the intersection of `TrackerRecoGeometryRecord` and `IdealMagneticFieldRecord`).

cmsbuild commented:
-1

**Failed Tests**: AddOn 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/summary.html
**COMMIT**: fa1e4e0f0de57c9d4afda51e577d47a2c19d4c7f
**CMSSW**: CMSSW_12_1_X_2021-09-13-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/35269/18601/install.sh` to create a dev area with all the needed externals and cmssw changes.

**CMS StaticAnalyzer warnings**: There are 10 EventSetupRecord::get warnings. See https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/llvm-analysis/esrget-sa.txt for details.

## AddOn Tests

- [hlt_mc_Fake](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/addOnTests/hlt_mc_Fake)
```
----- Begin Fatal Exception 14-Sep-2021 19:02:39 CEST-----------------------
An exception of category 'EventSetupConflict' occurred while
   [0] Calling beginJob
Exception Message:
two EventSetup Producers want to deliver type="MultipleScatteringParametrisationMaker" label=""
 from record NavigationSchoolRecord. The two providers are 
1) type="MultipleScatteringParametrisationMakerESProducer" label="hltMultipleScatteringParametrisationMakerESProducer"
2) type="MultipleScatteringParametrisationMakerESProducer" label="multipleScatteringParametrisationMakerESProducer"
Please either
   remove one of these Producers
   or find a way of configuring one of them so it does not deliver this data
   or use an es_prefer statement in the configuration to choose one.
----- End Fatal Exception -------------------------------------------------
```
- [hlt_mc_Fake1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/addOnTests/hlt_mc_Fake1)
```
----- Begin Fatal Exception 14-Sep-2021 19:02:45 CEST-----------------------
An exception of category 'EventSetupConflict' occurred while
   [0] Calling beginJob
Exception Message:
two EventSetup Producers want to deliver type="MultipleScatteringParametrisationMaker" label=""
 from record NavigationSchoolRecord. The two providers are 
1) type="MultipleScatteringParametrisationMakerESProducer" label="hltMultipleScatteringParametrisationMakerESProducer"
2) type="MultipleScatteringParametrisationMakerESProducer" label="multipleScatteringParametrisationMakerESProducer"
Please either
   remove one of these Producers
   or find a way of configuring one of them so it does not deliver this data
   or use an es_prefer statement in the configuration to choose one.
----- End Fatal Exception -------------------------------------------------
```
- [hlt_mc_Fake2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/addOnTests/hlt_mc_Fake2)
```
----- Begin Fatal Exception 14-Sep-2021 19:02:51 CEST-----------------------
An exception of category 'EventSetupConflict' occurred while
   [0] Calling beginJob
Exception Message:
two EventSetup Producers want to deliver type="MultipleScatteringParametrisationMaker" label=""
 from record NavigationSchoolRecord. The two providers are 
1) type="MultipleScatteringParametrisationMakerESProducer" label="hltMultipleScatteringParametrisationMakerESProducer"
2) type="MultipleScatteringParametrisationMakerESProducer" label="multipleScatteringParametrisationMakerESProducer"
Please either
   remove one of these Producers
   or find a way of configuring one of them so it does not deliver this data
   or use an es_prefer statement in the configuration to choose one.
----- End Fatal Exception -------------------------------------------------
```
<details>
<summary>Expand to see more addon errors ...</summary>

- [hlt_mc_GRun](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/addOnTests/hlt_mc_GRun)
- [hlt_mc_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/addOnTests/hlt_mc_HIon)
- [hlt_mc_PIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/addOnTests/hlt_mc_PIon)
- [hlt_mc_PRef](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/addOnTests/hlt_mc_PRef)
- [hlt_data_Fake](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/addOnTests/hlt_data_Fake)
- [hlt_data_Fake1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/addOnTests/hlt_data_Fake1)
- [hlt_data_Fake2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/addOnTests/hlt_data_Fake2)
- [hlt_data_GRun](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/addOnTests/hlt_data_GRun)
- [hlt_data_HIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/addOnTests/hlt_data_HIon)
- [hlt_data_PIon](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/addOnTests/hlt_data_PIon)
- [hlt_data_PRef](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18601/addOnTests/hlt_data_PRef)
</details>



## Comparison Summary



The workflows 140.53 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 1299 differences found in the comparisons
 - DQMHistoTests: Total files compared: 39
 - DQMHistoTests: Total histograms compared: 3001001
 - DQMHistoTests: Total failures: 3691
 - DQMHistoTests: Total nulls: 19
 - DQMHistoTests: Total successes: 2997269
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -45.703 KiB( 38 files compared)
 - DQMHistoSizes: changed ( 140.53 ): -44.531 KiB      Hcal/DigiRunHarvesting
 - DQMHistoSizes: changed ( 140.53 ): -1.172 KiB       RPC/DCSInfo
 - Checked 165 log files, 37 edm output root files, 39 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-09-13-2300+c6a8f8/45448/triggerResults): no differences found

makortel commented:
@Martin-Grunewald I see the addOn tests that run HLT and RECO in the same job. Should I then use the same label for the ESProducer in both HLT and RECO so those jobs would have only one instance of it? I.e. in `customizeHLTforCMSSW.py` just do `process.load('RecoTracker.TkMSParametrization.multipleScatteringParametrisationMakerESProducer_cfi')`?

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35269/25271

- This PR adds an extra 104KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HLTrigger/Configuration/python/customizeHLTforCMSSW.py  modified in PR(s): #34978, #34788
  - File  RecoTracker/IterativeTracking/python/MixedTripletStep_cff.py  modified in PR(s): #35194

cmsbuild commented:
Pull request #35269 was updated. @Martin-Grunewald, @jpata, @cmsbuild, @slava77 can you please check and sign again.

makortel commented:
> I.e. in `customizeHLTforCMSSW.py` just do `process.load('RecoTracker.TkMSParametrization.multipleScatteringParametrisationMakerESProducer_cfi')`?

At least that seemed to work in my local test.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35269/25272

- This PR adds an extra 104KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HLTrigger/Configuration/python/customizeHLTforCMSSW.py  modified in PR(s): #34978, #34788
  - File  RecoTracker/IterativeTracking/python/MixedTripletStep_cff.py  modified in PR(s): #35194

cmsbuild commented:
Pull request #35269 was updated. @Martin-Grunewald, @jpata, @cmsbuild, @slava77 can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18611/summary.html
**COMMIT**: 5b8d618033626a7463c06f246d2546b18a53a70d
**CMSSW**: CMSSW_12_1_X_2021-09-14-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/35269/18611/install.sh` to create a dev area with all the needed externals and cmssw changes.

**CMS StaticAnalyzer warnings**: There are 10 EventSetupRecord::get warnings. See https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18611/llvm-analysis/esrget-sa.txt for details.

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 39
 - DQMHistoTests: Total histograms compared: 3000833
 - DQMHistoTests: Total failures: 31
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3000779
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 38 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 165 log files, 37 edm output root files, 39 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-09-14-1100+c6a8f8/45458/triggerResults): no differences found

Martin-Grunewald commented:
@makortel 
Indeed, process load works best in such cases!

makortel commented:
Workflow 136.731 shows tiny differences in `Tracking/TrackParameters/generalTracks/SeedMon/detachedTripletStep/TrackBuilding`. The plots indicate one seed more in this iteration (assuming blue is reference and red is with PR). My best guess is some numerical change related to data being moved from thread local storage to "normal heap memory" (similar to going away from `static` in https://github.com/cms-sw/cmssw/pull/35081).

makortel commented:
test parameters
  - addpkg = DQM/Integration

makortel commented:
@cmsbuild, please test

makortel commented:
It came to my mind that the Online DQM clients are one place where an `import` could be missing.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18674/summary.html
**COMMIT**: 5b8d618033626a7463c06f246d2546b18a53a70d
**CMSSW**: CMSSW_12_1_X_2021-09-16-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/35269/18674/install.sh` to create a dev area with all the needed externals and cmssw changes.

**CMS StaticAnalyzer warnings**: There are 10 EventSetupRecord::get warnings. See https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18674/llvm-analysis/esrget-sa.txt for details.

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 39
 - DQMHistoTests: Total histograms compared: 3000833
 - DQMHistoTests: Total failures: 25
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3000785
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 38 files compared)
 - DQMHistoSizes: changed ( 312.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 165 log files, 37 edm output root files, 39 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-09-16-1100+c6a8f8/45516/triggerResults): no differences found

slava77 commented:
> This branch has conflicts that must be resolved
> Conflicting files
> HLTrigger/Configuration/python/customizeHLTforCMSSW.py

makortel commented:
> > This branch has conflicts that must be resolved
> > Conflicting files
> > HLTrigger/Configuration/python/customizeHLTforCMSSW.py

Will do in the next IB (apparently caused by https://github.com/cms-sw/cmssw/pull/35315).

slava77 commented on HLTrigger/Configuration/python/customizeHLTforCMSSW.py:
```diff
@@ -17,6 +17,10 @@
 
+# MultipleScatteringParametrisationMakerESProducer
+def customiseFor35269(process):
+    process.load("RecoTracker.TkMSParametrization.multipleScatteringParametrisationMakerESProducer_cfi")
```
I haven't seen any specific imports in the HLT configuration; in the configBuilder-based configurations the extra ES configs come via Configuration/StandardSequences
In this case CalibTracker/Configuration/python/Tracker_DependentRecords_forGlobalTag_nofakes_cff.py (included by Configuration/StandardSequences/python/AdditionalConditions_cff.py, included by the standard Configuration/StandardSequences/python/FrontierConditions_GlobalTag_cff.py) seems to be an appropriate place to insert

makortel replied:
The new ESProducer needs to eventually go in the HLT menu as well.

makortel replied:
Or does it actually?

makortel replied:
The ESProducer is needed in the HLT menu. Trying to run addOnTests without the customise function results in 
```
----- Begin Fatal Exception 20-Sep-2021 21:15:36 CEST-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 3
   [1] Running path 'HLT_IsoTrackHE_v4'
   [2] Calling method for module HitPairEDProducer/'hltPixelTracksHitDoublets'
Exception Message:
No data of type "MultipleScatteringParametrisationMaker" with label "" in record "NavigationSchoolRecord"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------
```
error in e.g. `hlt_mc_GRun` test.

slava77 replied:
I see, it's the addOnTests specifics. Got it.

Martin-Grunewald replied:
Yes, we'd add it to the HLT menu via CobfDB eventually. The customisation is just for this PR.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35269/25404

- This PR adds an extra 12KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HLTrigger/Configuration/python/customizeHLTforCMSSW.py  modified in PR(s): #34978, #35309, #35331
  - File  RecoTracker/IterativeTracking/python/MixedTripletStep_cff.py  modified in PR(s): #35194

cmsbuild commented:
Pull request #35269 was updated. @Martin-Grunewald, @jpata, @cmsbuild, @missirol, @slava77 can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18772/summary.html
**COMMIT**: 787488f812cce3e53a3b8dd45688929223713063
**CMSSW**: CMSSW_12_1_X_2021-09-20-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/35269/18772/install.sh` to create a dev area with all the needed externals and cmssw changes.

**CMS StaticAnalyzer warnings**: There are 10 EventSetupRecord::get warnings. See https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18772/llvm-analysis/esrget-sa.txt for details.

## Comparison Summary



The workflows 140.53 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 1299 differences found in the comparisons
 - DQMHistoTests: Total files compared: 40
 - DQMHistoTests: Total histograms compared: 3211080
 - DQMHistoTests: Total failures: 3691
 - DQMHistoTests: Total nulls: 19
 - DQMHistoTests: Total successes: 3207348
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -45.703 KiB( 39 files compared)
 - DQMHistoSizes: changed ( 140.53 ): -44.531 KiB      Hcal/DigiRunHarvesting
 - DQMHistoSizes: changed ( 140.53 ): -1.172 KiB       RPC/DCSInfo
 - Checked 169 log files, 37 edm output root files, 40 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-09-20-1100+c6a8f8/45608/triggerResults): no differences found

slava77 commented on RecoHI/HiTracking/python/HILowPtConformalPixelTracks_cfi.py:
```diff
@@ -2,6 +2,7 @@
 
 from RecoTracker.TkHitPairs.hitPairEDProducer_cfi import hitPairEDProducer as _hitPairEDProducer
 from RecoPixelVertexing.PixelTriplets.pixelTripletHLTEDProducer_cfi import pixelTripletHLTEDProducer as _pixelTripletHLTEDProducer
+from RecoTracker.TkMSParametrization.multipleScatteringParametrisationMakerESProducer_cfi import *
```
can this be instead included in `CalibTracker/Configuration/python/Tracker_DependentRecords_forGlobalTag_nofakes_cff.py` ?
@cms-sw/alca-l2 please clarify if there would be concerns about it.
I see that there we already include `RecoLocalTracker.SiStripRecHitConverter.StripCPEfromTrackAngle_cfi`, which has a superset of dependencies compared to MultipleScatteringParametrisationMakerESProducer.cc

makortel replied:
Fine for me if @cms-sw/alca-l2 agrees.

tvami replied:
We are a bit reluctant, to me it seems like `MultipleScatteringParametrisationMakerESProducer` is not really what belongs to `Tracker_DependentRecords_forGlobalTag_nofakes_cff`. Of course I understand that `StripCPEfromTrackAngle_cfi ` already includes it, but that's more a side-effect then the original purpose of that file. Can you please elaborate why you think it should be there instead of as it is now?

tvami replied:
I understand, tho I still don't see how `MultipleScatteringParametrisation` is a `tracker dependent record`. In the sense that the tracker material budget is influencing the multiple scattering?

slava77 replied:
>  I still don't see how `MultipleScatteringParametrisation` is a `tracker dependent record`. In the sense that the tracker material budget is influencing the multiple scattering?

yes, it's derived per `DetLayer` of the tracker; the relationship is built via `MultipleScatteringGeometry(const GeometricSearchTracker &`; the additional dependence on the magnetic field comes in through one of the methods that derives related set of layers (in this limited sense this hints to navigation).

tvami replied:
Ok Slava, I think this is convincing enough for us, no objections anymore from the AlCa side

slava77 reviewed (commented):
I think that it would be better to have a new record, smth like `MultipleScatteringRecord` or `TrackerMultipleScatteringRecord`, since the `NavigationSchoolRecord` does not seem to be intuitively named and does not serve a particularly related purpose.

Would it make sense to eventually move `MultipleScatteringGeometry` to ES as well? If so, it could go to the same MS record.

makortel commented on RecoTracker/TkTrackingRegions/src/RectangularEtaPhiTrackingRegion.cc:
```diff
@@ -126,7 +128,9 @@ std::unique_ptr<HitRZCompatibility> RectangularEtaPhiTrackingRegion::checkRZOld(
   if (theUseMS) {
-    MultipleScatteringParametrisation oSigma(layer, iSetup);
+    edm::ESHandle<MultipleScatteringParametrisationMaker> hmaker;
+    iSetup.get<NavigationSchoolRecord>().get(hmaker);
```
I'll remove all access to EventSetup from the TrackingRegion classes (by moving them to classes that create TrackingRegions) in the next PR.

makortel commented:
> I think that it would be better to have a new record, smth like `MultipleScatteringRecord` or `TrackerMultipleScatteringRecord`, since the `NavigationSchoolRecord` does not seem to be intuitively named and does not serve a particularly related purpose.

Ok, I can create a new Record (now that the approach itself appears to be acceptable). I'll probably go with `TrackerMultipleScatteringRecord` (it's anyway trivial to rename later if deemed useful).

> Would it make sense to eventually move `MultipleScatteringGeometry` to ES as well? If so, it could go to the same MS record.

It is used only when constructing `MSLayersKeeperX0AtEta` and `MSLayersKeeperX0Averaged`, and its constructor does not look very expensive (manipulations of vectors with size O(10)). My feeling is that while in principle the repetition of computations could be reduced, it would have negligible impact on timing (especially since it happens once per IOV).

makortel commented:
Changed the record to (a new) `TrackerMultipleScatteringRecord`, and moved the RECO-side imports of the cfi to `CalibTracker/Configuration/python/Tracker_DependentRecords_forGlobalTag_nofakes_cff.py`

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35269/25441

- This PR adds an extra 68KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HLTrigger/Configuration/python/customizeHLTforCMSSW.py  modified in PR(s): #35309, #35331

cmsbuild commented:
Pull request #35269 was updated. @malbouis, @yuanchao, @Martin-Grunewald, @missirol, @slava77, @jpata, @francescobrivio, @tvami can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18811/summary.html
**COMMIT**: 7cf94fc330d93ca2616e782c8dc85ba79dc5fe0c
**CMSSW**: CMSSW_12_1_X_2021-09-21-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/35269/18811/install.sh` to create a dev area with all the needed externals and cmssw changes.

**CMS StaticAnalyzer warnings**: There are 10 EventSetupRecord::get warnings. See https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6a8f8/18811/llvm-analysis/esrget-sa.txt for details.

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 40
 - DQMHistoTests: Total histograms compared: 3211080
 - DQMHistoTests: Total failures: 20
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3211038
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 39 files compared)
 - Checked 169 log files, 37 edm output root files, 40 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-09-21-2300+c6a8f8/45641/triggerResults): no differences found

slava77 commented:
+reconstruction

for https://github.com/cms-sw/cmssw/pull/35269	7cf94fc 
- code changes are somewhat technical in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no particularly relevant differences
    - as noted earlier https://github.com/cms-sw/cmssw/pull/35269#issuecomment-920311299 (and repeated in the following tests) there is one more seed in the detachedTripletStep, but it does not become a track, so the persisted data reco comparisons show no differences

tvami commented:
+alca

 * the only AlCa related change is the include of the multiple scattering parametrization in the tracker dependent cff

Martin-Grunewald commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

missirol commented:
@makortel I was having a look for HLT, and wanted to ask a clarification (maybe I'm missing something obvious).

IIuc, in https://github.com/cms-sw/cmssw/pull/35269#discussion_r712458334 you mention that `hlt_mc_GRun` (from the `addOnTests`) needs the HLT customisation function to run, but I understood from https://github.com/cms-sw/cmssw/pull/35269#issuecomment-919483900 that in that wf (and others) HLT and RECO run in the same job, and the new ES module would already be provided by RECO. What am I missing?

Maybe the HLT customisation is really needed, but not for that specific `addOnTest`?

makortel commented:
@missirol Looking closer, the step in `hlt_mc_GRun` that failed in https://github.com/cms-sw/cmssw/pull/35269#discussion_r712458334 was not the HLT+RECO job, but the following step that runs just the HLT menu. I tested with the current version of the PR commenting out the `customiseFor35269()`, and it continues to fail in similar way (while the RECO+HLT job runs fine because the ESProducer is provided via RECO, well now via `FrontierConditions_GlobalTag_cff`).

A consequence of the HLT+RECO jobs in the addOnTests that I referred to in https://github.com/cms-sw/cmssw/pull/35269#issuecomment-919483900 was the use of `process.load()` wrt. creating a second ESProducer with a different label (that would be technically possible, but would be more cumbersome to do).

missirol commented:
I see. All clear now. Thanks, Matti.

perrotta commented:
+1
